File: algorithms/652. Find Duplicate Subtrees/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def findDuplicateSubtrees(self, root):
        """
        :type root: TreeNode
        :rtype: List[TreeNode]
        """
        def serialize(node):
            if not node:
                return ""
            left = serialize(node.left)
            right = serialize(node.right)
            s = "%s,%s,%s" % (node.val, left, right)
            mem[s].append(node)
            return s
        mem = collections.defaultdict(list)
        serialize(root)
        return [nodes[0] for s, nodes in mem.iteritems() if len(nodes) > 1]

# 167 / 167 test cases passed.
# Status: Accepted
# Runtime: 105 ms
# beats 85.92 %

File: algorithms/652. Find Duplicate Subtrees/solution.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    unordered_map<string, vector<TreeNode *>> mem;
public:
    vector<TreeNode*> findDuplicateSubtrees(TreeNode* root) {
        serialize(root);
        vector<TreeNode *> result;
        for ( auto & p : mem )
            if ( p.second.size() > 1 )
                result.push_back(p.second[0]);
        return result;
    }

    string serialize(TreeNode * node) {
        if ( ! node )
            return "";
        string left = "," + serialize(node->left);
        string right = "," + serialize(node->right);
        string s = to_string(node->val) + left + right;
        mem[s].push_back(node);
        return s;
    }
};

// 167 / 167 test cases passed.
// Status: Accepted
// Runtime: 28 ms
// beats 82.61 %
